lib/salus/scanners/osv/gradle_osv.rb

```diff
@@ -0,0 +1,170 @@
+            #   "summary"=>"Privilege escalation", "details"=>"Vulnerability",
+            #   "aliases"=>["CVE-0000"], "modified"=>"2022-00-00T00:00:00.00Z",
+            #   "published"=>"2020-00-00T00:00:00Z", "references"=>[{"type"=>"ADVISORY",
+            #   "url"=>"https://nvd.nist.gov/vuln/detail/CVE-0000-0000"}],
+            #   "schema_version"=>"1.2.0", "severity"=>[{"type"=>"CVSS_V3",
+            #   "score"=>"CVSS:3.0"}], "database"=>"Github Advisory Database"
+            # }
+            match["ranges"].each do |version_ranges|
+              introduced, fixed = vulnerability_info_for(version_ranges)
+              if %w[SEMVER ECOSYSTEM].include?(version_ranges["type"]) &&
+                  version_matching(version, introduced, fixed)
+                puts "Found", match["package"]["name"]
+                results.append(format_vulnerability_result(match, version, introduced, fixed))
+              end
+            end
+          end
+        end
+      end
+      results
+    end
+
+    # Find dependencies from the project
+    def find_dependencies
+      shell_return = run_shell("/home/bin/parse_gradle_deps")
+      if !shell_return.success?
+        report_error(shell_return.stderr)
+        return []
+      end
+
+      begin
+        dependencies = JSON.parse(shell_return.stdout)
+      rescue JSON::ParserError
+        err_msg = "GradleOSV: Could not parse JSON returned by bin/parse_gradle_deps's stdout!"
+        report_stderr(err_msg)
+        report_error(err_msg)
+        return []
+      end
+
+      # Dedupe dependencies returned.
+      uniques = []
+      dependencies.group_by { |e| [e["group_id"], e["artifact_id"]] }.each do |_key, values|
+        uniques.append(values[0])
+      end
+
+      uniques
+    end
+
+    def format_vulnerability_result(match, version, introduced, fixed)
+      {
+        "Package": match.dig("package", "name"),
+        "Vulnerable Version": introduced,
+        "Version Detected": version,
+        "Patched Version": fixed,
+        "ID": match.fetch("aliases", [match.fetch("id", [])])[0],
+        "Database": match.fetch("database"),
+        "Summary": match.fetch("summary", match.dig("details")).strip,
+        "References": match.fetch("references", []).collect do |p|
+                        p["url"]
+                      end.join(", "),
+        "Source":  match.dig("database_specific", "url") || DEFAULT_SOURCE,
+        "Severity": match.dig("database_specific",
+                              "severity") || DEFAULT_SEVERITY
+      }
+    end
+
+    def vulnerability_info_for(version_range)
+      introduced = version_range["events"]&.first&.[]("introduced")
+      fixed = version_range["events"]&.[](1)&.[]("fixed")
+      [introduced.nil? ? EMPTY_STRING : introduced, fixed.nil? ? EMPTY_STRING : fixed]
+    end
+
+    # Fetch and Dedupe / Select Github Advisory over other sources when available.
+    def fetch_vulnerable_dependencies(dependencies)
```

Comment:
We have similar code in other scanners can we DRY this up and provide a small utility class for this feature?